File: OOPs/Java/Abstraction.java
package Java;

interface Car {
    void startEngine(); 
    void accelerate();
    void brake();
    void stopEngine();
}


class SportsCar implements Car {
    String brand;
    String model;
    boolean isEngineOn = false;
    int currentSpeed = 0;

    public SportsCar(String brand, String model) {
        this.brand = brand;
        this.model = model;
    }

    @Override
    public void startEngine() {
        isEngineOn = true;
        System.out.println(brand + " " + model + " : Engine starts with a roar!");
    }


    @Override
    public void accelerate() {
        if (!isEngineOn) {
            System.out.println(brand + " " + model + " : Engine is off! Cannot accelerate.");
            return;
        }
        currentSpeed += 20;
        System.out.println(brand + " " + model + " : Accelerating to " + currentSpeed + " km/h");
    }
    
    @Override
    public void brake() {
        currentSpeed -= 20;
        if (currentSpeed < 0) currentSpeed = 0;
        System.out.println(brand + " " + model + " : Braking! Speed is now " + currentSpeed + " km/h");
    }

    @Override
    public void stopEngine() {
        isEngineOn = false;
        currentSpeed = 0;
        System.out.println(brand + " " + model + " : Engine turned off.");
    }
}

public class Abstraction {
    public static void main(String[] args) {

        Car myCar = new SportsCar("Ford", "Mustang");

        myCar.startEngine();
        myCar.accelerate();
        myCar.accelerate();
        myCar.brake();
        myCar.stopEngine();

    }
}

File: OOPs/C++/Abstraction.cpp
#include<bits/stdc++.h>
using namespace std;

class Car {
    public:
    virtual void startCar() = 0; // Pure virtual function
    virtual void stopCar() = 0;
    virtual void accelerat() = 0;
    virtual void  brake() = 0;
    virtual ~Car(){};
};

class SportCar : public Car {

    private:
    string brand;
    string model;
    bool isEngineOn ;
    int speed ;

    public:

    SportCar(string brand,string model){
        this->brand = brand;
        this->model =  model;
        this->isEngineOn = false;
        this->speed  = 0;
    }

    void startCar(){
        if(!isEngineOn){
            isEngineOn = true;
            cout<<"Engine started for "<<brand<<" "<<model<<endl;
        }else{
            cout<<"Engine already started for "<<brand<<" "<<model<<endl;
        }
    }
    
    void stopCar(){
        if(isEngineOn){
            isEngineOn=false;
            speed=0;
            cout<<"Engine stop for "<<brand<<" "<<model<<endl;
        }else{
             cout<<"Engine already started for "<<brand<<" "<<model<<endl;
        }
    }

    void accelerat(){
        if(isEngineOn){
            speed+=10;
            cout<<"Accelerating "<<brand<<" "<<model<<". Current speed: "<<speed<<" km/h"<<endl;
        }else{
            cout<<"Cannot accelerate. Engineis off for  "<<brand<<" "<<model<<endl;
        }
    }

    void brake(){
        if(isEngineOn && speed > 0){
            speed -= 10;
            if(speed < 0) speed = 0; 
            cout<<"Braking "<<brand<<" "<<model<<". Current speed: "<<speed<<" km/h"<<endl;
        }else if(!isEngineOn){
            cout<<"Cannot brake. Engine is off for "<<brand<<" "<<model<<endl;
        }else{
            cout<<"Already at minimum speed for "<<brand<<" "<<model<<endl;
        }
    }

};

int main() {
    SportCar myCar("Ferrari", "488 GTB");
    
    myCar.startCar();
    myCar.accelerat();
    myCar.accelerat();
    myCar.brake();
    myCar.stopCar();

    return 0;
}